en/05_Contributing/05_Coding_Conventions/01_PHP_Coding_Conventions.md

```diff
@@ -6,190 +6,154 @@ iconBrand: php
+- Only provide docblock types for strongly typed parameters or return values if more context is provided.
 
-If you have to use raw SQL, make sure your code works across databases. Make sure you escape your queries like below,
-with the column or table name escaped with double quotes as below.
+## Mixed types
 
+- Where practical avoid using mixed types, including type unions, because this leads to complicated and brittle code.
+- Avoid using nullable scalars and instead use the default value of the scalar instead of null. One exception to this is for when `null` means 'uninitialised'.
+- Avoid using `func_get_args()`.
+
+## Object instantiation
+
+- For classes with the [`Injectable`](api:SilverStripe\Core\Injector\Injectable) trait, use Injector to instantiate objects using the `::create()` static method rather than the `new` keyword. This is done as it allows projects to swap out classes with different implementations.
+- For non-third-party classes without the [`Injectable`](api:SilverStripe\Core\Injector\Injectable) trait, consider adding the `Injectable` trait, or use the `new` keyword.
+- For third-party classes, [use `Injector` directly](/developer_guides/extending/injector/#basic-usage) to instantiate objects rather than the `new` keyword.
+- For unit tests, always use the `new` keyword when instantiating objects unless you are explicitly testing the injector logic.
+
+## Extensions and traits
+
+- Use the [`Extension`](api:SilverStripe\Core\Extension) class for extending classes, including DataObjects.
+- Do not use the DataExtension class, it will be deprecated in a future release.
+
+Use a trait instead of an [`Extension`](api:SilverStripe\Core\Extension) when the composable functionality:
+
+- Relies on per-instance properties, which cannot be added to extensions
+- Does not add or modify configuration properties or values
+- Is not expected to be applied to third-party code.
+
+Use an [`Extension`](api:SilverStripe\Core\Extension) instead of a trait when the composable functionality:
+
+- Does not rely on per-instance properties
+- Adds or modifies configuration properties or values
+- Could be expected to be applied to third-party code.
+
+### Extension hooks
+
+- For hooks that update a value before it's returned, name the hook `update{MethodName}` where `MethodName` has any prefix `get` removed. For example `getMyValue()` would have an extension hook named `updateMyValue()`.
+- For hooks that apply before or after another method is called within the method, name the hook `onBefore{MethodName}` or `onAfter{MethodName}` respectively. For instance if a method calls `processSomething()` call the hook `onBeforeProcessSomething()`.
+
+## Class Member Ordering
 
-```php
-MyClass::get()->where(['"Score" > ?' => 50]);
+Order code in classes in the following order:
 
-```
+- Class constants
+- Data-model definition static properties (`$db`, `$has_one`, `$many_many`, etc)
```

Comment:
```suggestion
- Configuration properties (i.e. private static properties that don't have the `@internal` annotation)
```
Seems like this is essentially what is meant here - and applies to more than just `DataObject` subclasses.

Reply:
Updated